src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,12 +421,148 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
+
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
+    {
+        callback(state->entry, ret);
+    }
+
+    free(state);
+}
+#endif
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+    return HAVE_GETADDRINFO_A;
+}
+
+// Get all address families and the canonical name
+static const struct addrinfo s_hostEntryForNameHints = {.ai_family = AF_UNSPEC, .ai_flags = AI_CANONNAME};
+
+int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo* info = NULL;
+
+    int result = getaddrinfo((const char*)address, NULL, &s_hostEntryForNameHints, &info);
+    if (result != 0)
+    {
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return GetHostEntries(address, info, entry);
+}
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
+    if (address == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    size_t addrlen = strlen((const char*)address);
+
+    if (addrlen > _POSIX_HOST_NAME_MAX)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct GetAddrInfoAsyncState* state = malloc(sizeof(*state) + addrlen + 1);
+
+    if (state == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_MEMORY;
+    }
+
+    memcpy(state->address, address, addrlen + 1);
+
+    *state = (struct GetAddrInfoAsyncState) {
+        .gai_request = {
+            .ar_name = state->address,
+            .ar_service = NULL,
+            .ar_request = &s_hostEntryForNameHints,
+            .ar_result = NULL
+        },
+        .gai_requests = &state->gai_request,
+        .sigevent = {
+            .sigev_notify = SIGEV_THREAD,
+            .sigev_value = {
+                .sival_ptr = state
+            },
+            .sigev_notify_function = GetHostEntryForNameAsyncComplete
+        },
+        .entry = entry,
+        .callback = callback
+    };
+
+    atomic_thread_fence(memory_order_release);
+
+    int32_t result = getaddrinfo_a(GAI_NOWAIT, &state->gai_requests, 1, &state->sigevent);
+
+    if (result != 0)
+    {
+        free(state);
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return result;
+#else
+    (void)address;
+    (void)entry;
+    (void)callback;
+
+    // GetHostEntryForNameAsync is not supported on this platform.
+    return -1;
+#endif
+}
+
 void SystemNative_FreeHostEntry(HostEntry* entry)
 {
     if (entry != NULL)
     {
-        free(entry->CanonicalName);
-        free(entry->IPAddressList);
+        if (entry->CanonicalName != NULL)
+        {
+            free(entry->CanonicalName);
+        }
+
+        if (entry->IPAddressList != NULL)
+        {
+            free(entry->IPAddressList);
+        }
```

Comment:
Nit: You don't need to null-check before calling free().  free() can take any value returned by malloc(), including NULL.  (Change is harmless, but something to keep in mind.)

Reply:
Thanks for the valuable input!

Maybe it was an artefact of the bug hunt, but Valgrind complained this. That's why I changed this. Will revert to the cleaner version.